config/typescript.js

```diff
@@ -1,20 +1,79 @@
 module.exports = {
   parser: '@typescript-eslint/parser',
+  parserOptions: {
+    'sourceType': 'module',
+  },
   plugins: [
     '@typescript-eslint',
   ],
-  extends: [
-    'eslint:recommended',
-    'plugin:@typescript-eslint/eslint-recommended',
-    'plugin:@typescript-eslint/recommended',
-  ],
   rules: {
+    // Checked by TypeScript - ts(2378)
+    'getter-return': 'off',
+    // Checked by TypeScript - ts(2300)
+    'no-dupe-args': 'off',
+    // Checked by TypeScript - ts(1117)
+    'no-dupe-keys': 'off',
+    // Checked by TypeScript - ts(7027)
+    'no-unreachable': 'off',
+    // Checked by TypeScript - ts(2367)
+    'valid-typeof': 'off',
+    // Checked by TypeScript - ts(2588)
+    'no-const-assign': 'off',
+    // Checked by TypeScript - ts(2588)
+    'no-new-symbol': 'off',
+    // Checked by TypeScript - ts(2376)
+    'no-this-before-super': 'off',
+    // Checked by TypeScript - `strictNullChecks`
+    'no-undef': 'off',
+    // Checked by TypeScript
+    'no-dupe-class-members': 'off',
+    // Checked by TypeScript
+    'no-redeclare': 'off',
+    // Checked by TypeScript
+    'import/no-unresolved': 'off',
+
     // Rules from @typescript-eslint/eslint-plugin
+    '@typescript-eslint/adjacent-overload-signatures': 'error',
     '@typescript-eslint/array-type': 'error',
-    '@typescript-eslint/consistent-type-definitions': ['error', 'type'],
+    '@typescript-eslint/ban-ts-ignore': 'error',
+    '@typescript-eslint/ban-types': 'error',
+    '@typescript-eslint/class-name-casing': 'error',
+    '@typescript-eslint/consistent-type-assertions': 'error',
+    '@typescript-eslint/consistent-type-definitions': ['error', 'interface'],
+    '@typescript-eslint/explicit-function-return-type': 'off',
+    '@typescript-eslint/explicit-member-accessibility': 'off',
+    '@typescript-eslint/explicit-module-boundary-types': 'off',
+    '@typescript-eslint/interface-name-prefix': 'error',
+    '@typescript-eslint/member-delimiter-style': ['error', {
+      'multiline': {
+        'delimiter': 'semi',
+        'requireLast': true,
+      },
+      'singleline': {
+        'delimiter': 'semi',
+        'requireLast': true,
+      },
+    }],
+    '@typescript-eslint/member-ordering': 'off',
+    '@typescript-eslint/no-dynamic-delete': 'off',
+    '@typescript-eslint/no-empty-interface': 'error',
     '@typescript-eslint/no-explicit-any': 'off',
+    '@typescript-eslint/no-extra-non-null-assertion': 'error',
+    '@typescript-eslint/no-extraneous-class': 'off',
+    '@typescript-eslint/no-inferrable-types': 'error',
+    '@typescript-eslint/no-misused-new': 'error',
+    '@typescript-eslint/no-namespace': ['error', { 'allowDefinitionFiles': true }],
+    '@typescript-eslint/no-non-null-asserted-optional-chain': 'error',
+    '@typescript-eslint/no-non-null-assertion': 'error',
+    '@typescript-eslint/no-parameter-properties': 'error',
+    '@typescript-eslint/no-require-imports': 'error',
```

Comment:
What do we do in cases of untyped JavaScript imports?

For instance, there are a lot of cases in `gaba` and elsewhere where `require` is used to import both `json-rpc-engine` and other libraries, made by us or others.

Reply:
If it's complicated, we can take that discussion offline. Worst case, I suppose we can disable this rule in the affected repos.

Reply:
I think this is the correct _default_ but yes, this doesn't work for a lot of code. We have two options, really:

1. Easy option is to add a ignore comment:

```ts
// eslint-disable-line @typescript-eslint/no-require-imports
const EthjsQuery = require('ethjs-query');
```

It's ugly but so it should be. 🤷‍♂

2. Import it the "correct way" and add a bare-bones type definition file.

```ts
import EthjsQuery from 'ethjs-query';
```

```ts
// project-root/@types/ethjs-query.d.ts or something, I'm writing this from memory
declare module 'ethjs-query';
```

Reply:
> both `json-rpc-engine` and other libraries, made by us or others.

The idea here is that this lint would get introduced as projects are "migrated", and where the migration is happening in dep tree order, libraries made by us should ship with their own type defns. The type defns that we write for libraries _made by others_ that aren't already on DefinitelyTyped can be shared across projects (more on that to come) but at worst each project could write their own barebones defns.

Reply:
> Worst case, I suppose we can disable this rule in the affected repos as a stopgap.

That too, this is but a default config.